### PR TITLE
MINOR: Refactor partition lag metric for cleaner encapsulation

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by hachikuji** [comment]
cc @becketqin Seemed a little unfortunate for `SubscriptionState` to need knowledge of partition lag metrics, so I took a shot at refactoring. Let me know if it seems reasonable.

**Comment by becketqin** [comment]
@hachikuji Thanks for the patch. I took a quick look and the Listener approach looks reasonable to me. I am currently traveling in China so my response might be delayed a little. I'll make a full pass tomorrow.

**Comment by hachikuji** [comment]
Thanks @becketqin. No hurry.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 1]
```diff
@@ -493,19 +491,19 @@ private long listOffset(TopicPartition partition, OffsetResetStrategy strategy)
...
-                        "position to {}", position, partitionRecords.partition, nextOffset);
+                    subscriptions.position(partitionRecords.partition, nextOffset);
+                }
 
-                subscriptions.position(partitionRecords.partition, nextOffset);
                 Long partitionLag = subscriptions.partitionLag(partitionRecords.partition);
-                if (partitionLag != null) {
-                    this.sensors.recordsFetchLag.record(partitionLag);
+                if (partitionLag != null)
                     this.sensors.recordPartitionFetchLag(partitionRecords.partition, partitionLag);
```
Comment (becketqin) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: Minor: Existing code,  would `recordPartitionLag` be a more accurate name as the lag is actually consume lag instead of fetch lag.

**Review by becketqin — COMMENTED** [review]
Thanks for the patch. LGTM except one minor comment.

**Comment by ijuma** [comment]
@hachikuji, I noticed that the KIP for the per partition metrics added tests to `BaseConsumerTest`. I think they don't need to run with all security protocols, so I suggest we move them to `PlaintextConsumerTest`. Maybe we can do that in this PR? Sorry, I missed that in my original review.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java [hunk 2]
```diff
@@ -473,4 +477,13 @@ private boolean isFetchable() {
 
     }
 
+    public interface Listener {
+        /**
+         * Fired after a new assignment is received (after a group rebalance or
+         * when the user manually changes the assignment).
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: Nit: unnecessary short lines?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 3]
```diff
@@ -480,9 +481,6 @@ private long listOffset(TopicPartition partition, OffsetResetStrategy strategy)
     }
 
     private List<ConsumerRecord<K, V>> drainRecords(PartitionRecords<K, V> partitionRecords, int maxRecords) {
-        if (partitionRecords.isDrained())
-            return Collections.emptyList();
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:484]: Just to double-check, this was removed because we want to update the metrics even if `isDrained` is true?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 4]
```diff
@@ -766,26 +764,13 @@ private void handleListOffsetResponse(Map<TopicPartition, Long> timestampsToSear
 
                 recordsCount = parsed.size();
 
-                if (!parsed.isEmpty()) {
-                    log.trace("Adding fetched record for partition {} with offset {} to buffered record list", tp, position);
-                    parsedRecords = new PartitionRecords<>(fetchOffset, tp, parsed);
-                }
+                log.trace("Adding fetched record for partition {} with offset {} to buffered record list", tp, position);
+                parsedRecords = new PartitionRecords<>(fetchOffset, tp, parsed);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:768]: Another question for my benefit: what is the reason for this change?

**Review by ijuma — COMMENTED** [review]
Nice clean-up, I left a couple of questions for my benefit.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 3 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:484]: That's right. I was trying to remove the need to handle empty fetch responses specially to update lag metrics. Before we simply discarded the fetch result if it had no records, but now we pass it through. Does this seem better or worse?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 4 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:768]: Related to my response above. We pass through empty fetches in order to ensure that the lag metric is updated.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 3 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:484]: Sounds good.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java (hunk 4 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:768]: Fair enough.

**Comment by hachikuji** [comment]
@ijuma Thanks for the review. I've moved the tests into `PlaintextConsumerTest`.

**Review by ijuma — APPROVED** [review]
LGTM

**Comment by guozhangwang** [comment]
Made a pass over the patch and cannot find a single line for comments :P LGTM!!

**Comment by ijuma** [comment]
Merged to trunk and 0.10.2 branches.
